### PR TITLE
Google hadoop file system tests

### PR DESCRIPTION
**Pull request by adam-stasiak**
Increased test coverage from 78% to 87 % in this class.

### CODE REVIEW

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 1]
```diff
@@ -514,6 +520,40 @@ public void createFile(Path filePath, byte[] data) throws IOException {
...
+    assertThrows("blockSize must be a positive integer: -1",IllegalArgumentException.class, () -> ghfs.create(path,true,-1,(short) 1,-1));
+  }
+
+  @Test
+  public void testRenameNullFile()  {
+    Path path = new Path("/directory1/");
+
+    assertThrows("src must not be null", NullPointerException.class, () -> ghfs.rename(null, null));
+    assertThrows("src must not be null", NullPointerException.class, () -> ghfs.rename(null, path));
+    assertThrows("dst must not be null", NullPointerException.class, () -> ghfs.rename(null, null));
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java]: This is a duplicate test case, also I think there are no value to have `(null, null)` test case, if there are already `(path, null)` and `(null, path)` test cases.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 2]
```diff
@@ -514,6 +520,40 @@ public void createFile(Path filePath, byte[] data) throws IOException {
...
+    assertThrows("dst must not be null", NullPointerException.class, () -> ghfs.rename(null, path));
+  }
+
+  @Test
+  public void testListStatusNull()  {
+    @Nullable
+    Path path = null;
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java]: Just pass `null` as a parameter without variable (here and below)

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 3]
```diff
@@ -573,6 +613,39 @@ public void testGlobStatus() throws IOException {
...
+  public void testGlobStatusOptions() throws IOException {
+    testGlobStatusFlatConcurrent(true,true);
+    testGlobStatusFlatConcurrent(true,false);
+    testGlobStatusFlatConcurrent(false,true);
+    testGlobStatusFlatConcurrent(false,false);
+  }
+
+  private void testGlobStatusFlatConcurrent(boolean flat, boolean concurent) throws IOException {
+    Configuration configuration = getConfigurationWtihImplementation();
+    configuration.set(GoogleHadoopFileSystemConfiguration.GCS_FLAT_GLOB_ENABLE.getKey(), String.valueOf(flat));
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java]: Use `setBoolean()` method instead of `set()`

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 4]
```diff
@@ -573,6 +613,39 @@ public void testGlobStatus() throws IOException {
...
+    Path testRoot = new Path("/directory1/");
+    ghfs.mkdirs(testRoot);
+    ghfs.mkdirs(new Path("/directory1/subdirectory1"));
+
+    byte[] data = new byte[10];
+    for (int i = 0; i < data.length; i++) {
+      data[i] = (byte) i;
+    }
+
+    createFile(new Path("/directory1/subdirectory1/file1"), data);
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java]: You can just convert string to byte array, instead of creating it manually:
`"data".getBytes(StandardCharsets.UTF_8)`

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 5]
```diff
@@ -573,6 +613,39 @@ public void testGlobStatus() throws IOException {
...
+
+    FileStatus[] rootDirectories = ghfs.globStatus(new Path("/d*"));
+    assertThat(rootDirectories).hasLength(1);
+    assertThat(rootDirectories[0].getPath().getName()).isEqualTo("directory1");
+
+    ghfs.delete(testRoot, true);
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java]: add `recursive=` comment before, `true` value:
`, /* recursive= */ true)`

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 6]
```diff
@@ -650,8 +723,23 @@ private void testFileChecksum(
     assertThat(fileChecksum.getAlgorithmName()).isEqualTo(checksumType.getAlgorithmName());
     assertThat(fileChecksum.getLength()).isEqualTo(checksumType.getByteLength());
     assertThat(fileChecksum.getBytes()).isEqualTo(checksumFn.apply(fileContent));
+    assertThat(fileChecksum.toString()).containsMatch(String.format("%s: *",checksumType.getAlgorithmName()));
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java]: replace with `.contains(checksumType.getAlgorithmName())` to be less confusing?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 7]
```diff
@@ -650,8 +723,23 @@ private void testFileChecksum(
...
     assertThat(ghfs.delete(filePath, true)).isTrue();
   }
+
+  @Test
+  public void testInitializeWithEmptyWorkingDirectory() throws IOException {
+    GoogleHadoopFileSystem myGhfs = (GoogleHadoopFileSystem) ghfs;
+    Configuration config = loadConfig();
+    ghfs.initialize(myGhfs.initUri, config);
+    String rootBucketName = myGhfs.getRootBucketName();
+    config.set(
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java]: Use `unset()` method instead of setting to empty string?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemNewUriFormatIntegrationTest.java [hunk 8]
```diff
@@ -237,5 +237,96 @@ public void testGetGcsPath() throws URISyntaxException {
...
+
+  @Test
+  public void testGcsLazyConfigurationEnabled() throws IOException, URISyntaxException {
+    GoogleHadoopFileSystem testInstance = new GoogleHadoopFileSystem();
+    ghfs = testInstance;
+    ghfsFileSystemDescriptor = testInstance;
+
+    ghfsHelper = new HadoopFileSystemIntegrationHelper(ghfs, ghfsFileSystemDescriptor);
+    Configuration conf = loadConfig();
+    conf.set(
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemNewUriFormatIntegrationTest.java]: Use `.setBoolean()` method.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemNewUriFormatIntegrationTest.java [hunk 9]
```diff
@@ -47,40 +49,40 @@
...
-
-          HadoopFileSystemTestBase.postCreateInit();
-        }
-
-        @Override
-        public void after() {
-          GoogleHadoopFileSystemTestBase.storageResource.after();
-        }
-      };
+          new NotInheritableExternalResource(GoogleHadoopFileSystemNewUriFormatIntegrationTest.class) {
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemNewUriFormatIntegrationTest.java]: Please, do not format lines that are not changed, only changed ones:
```bash
git diff -U0 master | python ~/dev/src/google-java-format/scripts/google-java-format-diff.py -p1 -i
```

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java (hunk 1 above)
Comment (adam-stasiak) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java]: this scenario is only for having 100% branches coverage.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 10]
```diff
@@ -651,8 +649,91 @@ private void testFileChecksum(
...
+    String rootBucketName = myGhfs.getRootBucketName();
+    config.unset(GoogleHadoopFileSystemConfiguration.GCS_WORKING_DIRECTORY.getKey());
+    ghfs.initialize(myGhfs.initUri, config);
+
+    assertThat(ghfs.getHomeDirectory().toString()).startsWith("gs://" + rootBucketName);
+  }
+
+  @Test
+  public void testGlobStatusOptions() throws IOException {
+    testGlobStatusFlatConcurrent(true, true);
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java]: Add parameter name comments (`/* <par>= */`) before `true` and `false` values.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 11]
```diff
@@ -651,8 +649,91 @@ private void testFileChecksum(
...
     // Cleanup.
-    assertThat(ghfs.delete(filePath, true)).isTrue();
+    assertThat(ghfs.delete(filePath, /* recursive= */ true)).isTrue();
+  }
+
+  @Test
+  public void testInitializeWithEmptyWorkingDirectory() throws IOException {
+    GoogleHadoopFileSystem myGhfs = (GoogleHadoopFileSystem) ghfs;
+    Configuration config = loadConfig();
+    ghfs.initialize(myGhfs.initUri, config);
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java]: Is this initialize necessary? Seems like there are another one below.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 12]
```diff
@@ -651,8 +649,91 @@ private void testFileChecksum(
...
     assertThat(fileChecksum.getBytes()).isEqualTo(checksumFn.apply(fileContent));
+    assertThat(fileChecksum.toString()).contains("%s: ".format(checksumType.getAlgorithmName()));
 
     // Cleanup.
-    assertThat(ghfs.delete(filePath, true)).isTrue();
+    assertThat(ghfs.delete(filePath, /* recursive= */ true)).isTrue();
+  }
+
+  @Test
+  public void testInitializeWithEmptyWorkingDirectory() throws IOException {
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java]: Add description of the expected result to test name: `testInitializeWithEmptyWorkingDirectory_should...`

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 13]
```diff
@@ -651,8 +649,91 @@ private void testFileChecksum(
...
+
+    ghfs.initialize(ghfs.getUri(), configuration);
+    Path testRoot = new Path("/directory1/");
+    ghfs.mkdirs(testRoot);
+    ghfs.mkdirs(new Path("/directory1/subdirectory1"));
+    createFile(
+        new Path("/directory1/subdirectory1/file1"), "data".getBytes(StandardCharsets.UTF_8));
+    FileStatus[] rootDirectories = ghfs.globStatus(new Path("/d*"));
+
+    assertThat(rootDirectories).hasLength(1);
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java]: No need to check number of records, if exact content already asserted in next assertion.

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemTest.java [hunk 14]
```diff
@@ -143,6 +145,42 @@ public MethodOutcome renameFileIntoRootOutcome() {
         });
   }
 
+  @Test
+  public void testCouldUseFlatGlob() throws IOException, URISyntaxException {
+    Configuration lazyConf = new Configuration();
+    lazyConf.set(GCS_LAZY_INITIALIZATION_ENABLE.getKey(), "true");
```
Comment (medb) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemTest.java]: `setBoolean`

**Comment by adam-stasiak** [comment]
I rebased this pull request to be up to date with master. I think all suggestions are applied. @medb I am not sure - should this test class be migrated to the new way of tests initialization introduced by mgasiorowski?

**Comment by adam-stasiak** [comment]
Ready for another review
